src/foam/dao/SyncDAO.js

```diff
@@ -16,248 +16,391 @@
-        a = a.array;
-        for ( var i = 0 ; i < a.length ; i++ ) {
-          this.remove_(x, a[i]);
-        }
-      }.bind(this));
-    },
-
-    /** @private */
-    function processFromServer(obj) {
-      this.delegate.put(obj).then(function(obj) {
-        this.syncRecordDAO.put(
-          this.SyncRecord.create({
-            id: obj.id,
-            syncNo: obj[this.syncProperty.name]
-          }));
-      }.bind(this));
-    },
-
-    /** @private */
-    function syncFromServer() {
-      var E = this;
-
-      this.syncRecordDAO.select(E.MAX(this.SyncRecord.SYNC_NO)).then(function(m) {
-        this.remoteDAO
-          .where(E.GT(this.syncProperty, m.value || 0))
-          .select().then(function(a) {
+      // Marks all the removed objects' sync records as deleted via remove_().
+      return this.delegate.select_(x, null, skip, limit, order, predicate).
+          then(function(a) {
             a = a.array;
+            var promises = [];
+
             for ( var i = 0 ; i < a.length ; i++ ) {
-              this.processFromServer(a[i]);
+              promises.push(this.remove_(x, a[i]));
             }
+
+            return Promise.all(promises);
           }.bind(this));
-      }.bind(this));
     },
 
-    /** @private */
-    function syncToServer() {
-      var E = this;
-      var self = this;
+    //
+    // Private synchronization details.
+    //
 
-      this.syncRecordDAO
-        .where(E.EQ(this.SyncRecord.SYNC_NO, -1))
-        .select().then(function(records) {
-          records = records.array;
+    {
+      name: 'putFromRemote_',
+      documentation: 'Process a put() to cache from remote.',
+      code: function(obj) {
+        // console.log('putFromRemote_', foam.json.stringify(obj));
+        var self = this;
+        var ret;
+        return self.withSyncRecordTx_(function() {
+          return self.delegate.put(obj).then(function(o) {
+            ret = o;
+            // console.log('putFromRemote_: store syncRecord for', foam.json.stringify(o));
+            return self.syncRecordDAO.put(self.SyncRecord.create({
+              id: o.id,
+              syncNo: o[self.versionProperty.name]
+            }));
+          });
+        }).then(function() { return ret; });
+      }
+    },
+    {
+      name: 'removeFromRemote_',
+      documentation: 'Process a remove() on cache from remote.',
+      code: function(obj) {
+        // console.log('removeFromRemote_', foam.json.stringify(obj));
+        var self = this;
+        var ret;
+        return self.withSyncRecordTx_(function() {
+          return self.delegate.remove(obj).then(function(o) {
+            ret = o;
+            // console.log('removeFromRemote_: store syncRecord for', foam.json.stringify(obj));
+            return self.syncRecordDAO.put(self.SyncRecord.create({
+              id: obj.id,
+              syncNo: obj[self.versionProperty.name],
+              deleted: true
```

Comment:
I think what you propose would probably be okay. I'd have to run it through the stress test a few times to be more confident. One down side of that approach is that, until we see another non-deletion update, we will re-process this deletion on each sync in the polling case. Since we push our updates first, that shouldn't be a problem, but as I say, I want to think about it some more and run it through the stress test a few times before implementing the change.